### PR TITLE
Add support for auto packing ratio

### PR DESCRIPTION
**Pull request by irenedea**
Adds 'auto' packing ratio for finetuning.

If 'auto' is specified, packing will be profiled for the dataloader configuration across a maximum of 20 packing ratios until the highest packing ratio with 0 waste is found.

If there are multiple ranks, we take the minimum 'auto' packing ratio across all ranks.

## Testing
- Manual tests
     -  1 epoch on mosaicml/dolly_hhrlhf, auto packing had 1.3% waste overall
         - finetune-auto-pack-test-CHlwSQ https://wandb.ai/mosaic-ml/irene-test/runs/7ccl7k7a 
             - time/token: 9390338, time/sample: 4896, time/batch: 51, time/epoch: 1
         - finetune-auto-pack-test-baseline2-zptNpN https://wandb.ai/mosaic-ml/irene-test/runs/b0xqxmlj
              - time/token: 9518016, time/sample: 59288, time/batch: 618, time/epoch: 1
     - No packing run are the same before and after PR 
              - before: finetune-auto-pack-test-baseline2-zptNpN , after: finetune-ratio-no-pack-branch-test-8N5I1y
     - Packing ratio of 3.0 is same before and after PR
              - before: finetune-ratio-3-pack-test-kW8gp6 , after: finetune-ratio-3-pack-with-auto-test-YVP00E 
- Unit tests
     - Tests packing on small tensors
     - Tests packing on small tensors with leftovers and waste calculation
     - Tests auto packing ratio selection for single and multi ranks
     - Tests auto packing with dataloader 
     - Adds 'auto' packing_ratio parameterization to existing dataloader tests

### CODE REVIEW

File: llmfoundry/data/packing.py
Comment (irenedea) [inline llmfoundry/data/packing.py:275]: This assumes ~100 tokens for the "smallest" examples. Would like to get some feedback on whether or not this is reasonable.

future optimization: profile for this as well?

File: llmfoundry/data/dataloader.py
Comment (irenedea) [inline llmfoundry/data/dataloader.py:17]: refactored here to share this code because we were essentially copying the same code in packing.py

File: llmfoundry/data/packing.py
Comment (irenedea) [inline llmfoundry/data/packing.py:327]: Set this to 0 workers because [a single worker is responsible for a single batch](https://discuss.pytorch.org/t/guidelines-for-assigning-num-workers-to-dataloader/813). Because we load one big batch only, there's no need for multiple workers. (Additionally, having more than one worker here caused the cpu tests to fail in CI, but I couldn't repro in interactive.)

Alternatively instead of just loading a single big batch, we could use the dataloader with multiple workers to get multiple smaller batches.

File: llmfoundry/data/packing.py
Comment (alextrott16) [reply llmfoundry/data/packing.py:277]: Yeah, I agree that this is probably a pretty good heuristic for getting an upper bound on the packing ratio.

If there's an issue, it's probably that with this heuristic, `num_packing_ratios = 20` might not have enough granularity. It might be worth coming up with a less brute force approach that better guarantees that you'll get fine-grained search in the most relevant range of packing ratios.

Off the top of my head, one idea is to start from 1, and double packing ratio until you hit one that is too high (has waste), then next look at half-way between the highest safe packing ratio you've checked and the lowest >0 waste packing ratio you've checked. Do that until the half-way point is rounded into a `raw_batch_size` that you've already checked. Then pick the best packing ratio from the set you've covered.

File: scripts/train/finetune_example/mpt-7b-arc-easy--gpu.yaml
Comment (alextrott16) [inline scripts/train/finetune_example/mpt-7b-arc-easy--gpu.yaml:44]: Should we instead update the YAMLs to recommend the 'auto' option?

File: tests/test_dataloader.py
Comment (alextrott16) [inline tests/test_dataloader.py:255]: Nit (sorry): Should the signature be `Optional[Union[int, float, str]]`?

**Review by alextrott16 — COMMENTED** [review]
LGTM. Thanks for adding this!!
Will hold off on approval so this can get another set of eyes!

Also left a couple comments (nothing major) along with a suggestion for an alternative way to search for the optimal packing ratio. I think re-using my brute force approach might not be the best way to go. If you agree, that will require more of an overhaul of this code, unfortunately. But potentially better to do that than simply inherit my hasty decision :)

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:60]: ah yes that's right! thanks for catching that

File: tests/test_dataloader.py
Comment (irenedea) [reply tests/test_dataloader.py:255]: good point! i think it would be `Optional[Union[float, str]]`?

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:60]: hmmm I think the type isn't necessarily consistent because during profiling the type of examples is `Dict[str, torch.Tensor]` while self.base_collator is `lambda x: x`

Comment (irenedea) [reply llmfoundry/data/packing.py:60]: maybe we can refactor this out into two functions. I think that'll be cleaner also:
```
def __call__(self, examples: List[Dict[str, torch.Tensor]]):
    batch = self.base_collator(examples)
    return self.pack(batch)

def pack(self, examples: Dict[str, torch.Tensor]):
    # assertions, packing logic
```
and then in profiling we can just called BinPackCollator.pack

File: llmfoundry/data/packing.py
Comment (irenedea) [inline llmfoundry/data/packing.py:362]: @alextrott16 Do you think we should be passing in dataloader.collate_fn here instead of `lambda x: x` ?

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:279]: Thanks for the insight! Getting better granularity sounds like a good next step. 
Another idea: If we feel good about the the max heuristic we could also do a binary search that starts between the min and max.

Separately, I think it would be appropriate to do this as a follow up PR since this PR already has a lot of other refactors and new tests. What do you think? @alextrott16 @dakinggg

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py]: Updated to do this refactor. (It was needed to appease the pyright gods anyway)

File: llmfoundry/data/packing.py
Comment (alextrott16) [reply llmfoundry/data/packing.py:370]: I don't think we need to, unless you prefer to refactor the surrounding code. The dataloader that is set up in this script already applies the seq2seq collator function, so, in the context of this script, having `BinPackCollator` wrap an identify function should be correct in the context of the current logic.

File: llmfoundry/data/packing.py
Comment (alextrott16) [reply llmfoundry/data/packing.py]: Sorry for not chiming in sooner. I'm on board with any refactor that makes sense to you and whatever aligns this code with the auto packing script such that pyright is appeased.

File: llmfoundry/data/packing.py
Comment (alextrott16) [reply llmfoundry/data/packing.py:279]: Fine with me if it's a separate PR, but I would advocate for it being a fast follow-up since the 'auto' implementation will potentially change the packing ratio that it comes up with.

File: llmfoundry/data/packing.py
Comment (dakinggg) [inline llmfoundry/data/packing.py:293]: I think the way to do this generically is

```
if dist.is_available() and dist.is_initialized():
    device = get_device(None)
    packing_ratio_tensor = device.tensor_to_device(torch.tensor(packing_ratio))
    dist.all_reduce(packing_ratio_tensor, reduce_operation='MIN')
    packing_ratio = packing_ratio_tensor.item()
```

File: llmfoundry/data/packing.py
Comment (dakinggg) [inline llmfoundry/data/packing.py:287]: are the profiling results already sorted?

File: llmfoundry/data/packing.py
Comment (dakinggg) [inline llmfoundry/data/packing.py:348]: This controls how many samples we are going to use for profiling, yeah? So this says we will profile `device_batch_size * max_ratio * 100 samples`. I think you may have already run this, but could you comment here for long this takes to run, and how changing this to 200, 300, 400 changes how long it takes to run? Would help to make an informed decision about this number.

**Review by dakinggg — COMMENTED** [review]
Looking good! 

Two more manual tests I would like to see before merging:
(1) a short training run without packing, before and after this PR. These should be identical.
(2) a short training run with a set packing ratio, before and after this PR. These should be identical.

File: llmfoundry/data/dataloader.py
Comment (irenedea) [reply llmfoundry/data/dataloader.py:15]: circuilar dependencies, iirc it was llmfoundry/data is dependent on llmfoundry/utils and vice versa (if I added this in builders.py)

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:287]: yes! i'll include a comment about this

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:293]: when I use dist without cuda I get errors--is it supposed to work with other backends? (I think yes, but I haven't gotten it to work...)

File: scripts/misc/profile_packing.py
Comment (irenedea) [reply scripts/misc/profile_packing.py:77]: the max number of "leftover" examples that didn't get packed. They get kept and the algorithm can try again to fit them. Right now it defaults to 10 * device batch size . Should we default to keeping all of them for auto to try to reduce waste as much as possible?  cc @alextrott16

File: tests/test_dataloader.py
Comment (irenedea) [reply tests/test_dataloader.py:254]: It can be, but I think for typing purposes float also accepts int (even though it's not a proper subclass)

Comment (irenedea) [reply tests/test_dataloader.py:254]: will use `Literal['auto']`, that's better than just `str`

File: llmfoundry/data/packing.py
Comment (dakinggg) [reply llmfoundry/data/packing.py:293]: yes, it should work with gloo backend as well. but this is not a big deal to investigate. can you make a jira to fix this and put it in the llm foundry backlog epic?

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:348]: I think we should keep it at 100 because it scales linearly and with 100, we're looking at about several minutes to profile. (Based on profiling on dolly)

Comment (irenedea) [reply llmfoundry/data/packing.py:348]: If anything I think we could optimize by using the dataloader with more workers etc and not get a single large batch but mutiple smaller batches. I can create a jira to investigate optimizations here

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:279]: done

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:293]: actually nvm, i'm not sure what i was doing before, but i tested it out again with gloo just now and it works fine. I suspect I was running nccl x cpu previously. 

```
from composer.utils import dist, get_device
import torch

torch.distributed.init_process_group('gloo')

if dist.is_available() and dist.is_initialized(): 
    device = get_device(None) 
    if dist.get_local_rank() == 0:
        packing_ratio_tensor = device.tensor_to_device(torch.tensor(5.0)) 
    else:
        packing_ratio_tensor = device.tensor_to_device(torch.tensor(4.0)) 

    dist.all_reduce(packing_ratio_tensor, reduce_operation='MIN') 
    packing_ratio = packing_ratio_tensor.item()
    print(packing_ratio)
```

Comment (irenedea) [reply llmfoundry/data/packing.py:293]: will update to generic

File: llmfoundry/data/packing.py
Comment (dakinggg) [reply llmfoundry/data/packing.py:368]: hmm ok. sounds fine for now. please make the jira for further optimization and put it in foundry backlog epic

File: scripts/misc/profile_packing.py
Comment (dakinggg) [reply scripts/misc/profile_packing.py:77]: yeah, lets keep em all

Comment (dakinggg) [reply scripts/misc/profile_packing.py:77]: but also, my question was more that the variable doesn't seem to be used anywhere?

File: llmfoundry/data/packing.py
Comment (dakinggg) [inline llmfoundry/data/packing.py:370]: I'm contemplating some determinism stuff here. What happens if we change the number of nodes...? I think you could end up with a different packing ratio, right? Should we turn off the distributed sampler for the profiling so that it is deterministic? This would negate any profiling benefit of having many processes, but maybe we should be on the safe side for now?

If we do this, please redo the profiling run that tested waste percent, in case this is a really bad decision :)

**Review by dakinggg — COMMENTED** [review]
LGTM except for one determinism related comment

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:370]: hmmm yeah, also:
if you change the number of nodes without adjusting global batch size, then device batch size will decrease. This will also affect the profiling results...

Comment (irenedea) [reply llmfoundry/data/packing.py:370]: Is this accurate:
- changing the device batch size can result in a different packing ratio
- changing global batch size can result in resuming at the incorrect point in the dataset

Comment (irenedea) [reply llmfoundry/data/packing.py:370]: resolved offline, we don't need to support elastic resumption for now

File: llmfoundry/data/packing.py
Comment (dakinggg) [inline llmfoundry/data/packing.py:293]: should we set this to a different value depending on rank so that each rank profiles different data?

**Review by dakinggg — APPROVED** [review]
LGTM, with one more question relating to the randomness

File: llmfoundry/data/packing.py
Comment (irenedea) [reply llmfoundry/data/packing.py:293]: I think the distributed sampler/streaming dataset that gets created with the dataloader should handle this case. Is there something I am missing?

File: scripts/misc/profile_packing.py
Comment (irenedea) [reply scripts/misc/profile_packing.py:77]: oh yup, good catch. It's not necessary to pop it here

File: llmfoundry/data/packing.py
Comment (dakinggg) [reply llmfoundry/data/packing.py:290]: oh yeah, i confused myself. all good.

**Pull request merged**
